### DISCUSSION TITLE
Allow marking a module "available for parallel compilation", and compile all of the scala sources in parallel

### DISCUSSION THREAD

**Discussion by haemin-tokenz** [com-lihaoyi/mill#3474] [Ideas]
In my work project, I have a lot of modules that I've structured like the following: `xxx-interface` and `xxx-impl`. 

The implementation module contains a lot of code that are simply implementations of the interface, and as such, between the files inside `xxx-impl`, there are no dependencies. Naturally the implementation module has much more code and is slower to compile.

If there is a way to mark a module as "parallel compilation", such that each and one of those files are invisible to each other so we can compile all of them in parallel, it would be a game changer for me. 

* The compilation result should remain the same
* Referring to other files should result in a not found error at compile time (bonus points to manually allow some files to depend on some specific other files)

**Comment by lihaoyi** [comment]
There currently isn't such a way, but maybe one could be developed. @bishabosha did some work on parallelizing compilation in Scala 3, and there's some prior work with Hydra in Scala 2, but Mill doesn't have anything built in. We use the same compilation toolchain as the rest of the Scala ecosystem

**Reply by haemin-tokenz** [reply]
Hey thanks for the reply. This is not about parallelizing the compiler, but rather treat each files as a separate project (in the compilation unit sense) and aggregate the output.

**Reply by lihaoyi** [reply]
If you want to do that, maybe you can use Cross Builds? Something like (warning, not tested)


```scala
object wrapper extends ScalaModule{
  def scalaVersion = "2.13.14"
  val submoduleFiles = os.walk(millSourcePath).filter(_.ext == "scala").map(
  object file extends Cross[CrossFile](submoduleFiles.map(_.relativeTo(wrapper.millSourcePath)))
  trait CrossFile extends ScalaModule with Cross.Value[String] {
    def scalaVersion = "2.13.14"
    def allSourceFiles = Agg(PathRef(wrapper.millSourcePath / crossValue))
  }
  def moduleDeps = file.crossModules
}
```

That would give you a bunch of `wrapper.file[Foo.scala]`, `wrapper.file[Bar.scala]`, etc. submodules that each have one source file, and then `wrapper` is a Scala module that depends on all of them so you can depend on `wrapper` to treat them as one big unit but individually they all compile in parallel

**Reply by haemin-tokenz** [reply]
Amazing. Thanks for your input, let me see if I can get it to work.

**Reply by lihaoyi** [reply]
@kaepaokaawela let me know if it works, or if it doesn't. This may come in useful for others in future

**Reply by haemin-tokenz** [reply]
Sure. I'm quite busy with other things right now, but I'll report back. 

I'm expecting a week or two until I have time to take a look at this...

**Reply by lolgab** [reply]
The following seems to work.

```scala
import mill._
import mill.scalalib._

object `package` extends RootModule with ScalaModule {
  def scalaVersion = "3.5.0"

  val fileSrc = millSourcePath / "file" / "src"
  def fileSrcInput = T.input { PathRef(fileSrc) }
  interp.watch(fileSrc)
  val submoduleFiles = os
    .walk(fileSrc)
    .filter(f => f.ext == "scala")
    .map(_.relativeTo(fileSrc))
  object file extends Cross[CrossFile](submoduleFiles.map(_.toString.replace("/", "_")))
  trait CrossFile extends ScalaModule with Cross.Module[String] {
    def scalaVersion = "2.13.14"
    def allSourceFiles = T {
      val dir = fileSrcInput().path / crossValue
      submoduleFiles.map(f => PathRef(fileSrc / f))
    }
    def sources = Seq.empty[PathRef]
  }

  def moduleDeps = file.crossModules
}
```

I think that this idea can be improved by introducing a dynamic tree of dependencies..
Given a tree of directories, you can create a module for every directory, where all files belong to the module and subdirectories are submodules. You can make the parent directory module depend on the child.
This can be great for code organization, since it makes it impossible from a deeper node to depend on a higher node on the tree, making it impossible to create a spaghetti code of dependencies.
It could be a plugin or a built-in trait to be added to Mill.

**Reply by lolgab** [reply]
I implemented a prototype of the `TreeModule` which applies dependencies is tree structure:
```scala
import mill._
import mill.scalalib._

/* Your build code */
trait Common extends ScalaModule {
  def scalaVersion = "3.5.0"
}
object module extends Common with TreeModule {
  object cross extends Cross[Shared](crossValues)
  trait Shared extends Common with TreeCross {
    def ivyDeps = super.ivyDeps() ++ Agg(
      ivy"com.lihaoyi::os-lib::0.10.6"
    )
  }
}

/* Library code */
trait TreeModule extends ScalaModule { outer =>
  private def fileSrc = millSourcePath / "src"
  os.makeDir.all(fileSrc)
  interp.watch(fileSrc)

  private def fileSrcInput = T.input { PathRef(fileSrc) }
  private def mods(paths: Seq[os.Path]) =
    paths
      .filter(os.isDir(_))
      .map(_.subRelativeTo(fileSrc))

  val crossValues = mods(os.walk(fileSrc))

  trait TreeCross extends ScalaModule with Cross.Module[os.SubPath] {
    override def millSourcePath = super.millSourcePath / os.up
    def allSourceFiles = T {
      val dir = fileSrcInput().path / crossValue
      val res =
        if (os.isDir(dir))
          os.list(dir)
            .filter(f => f.ext == "scala" || f.ext == "java")
            .map(PathRef(_))
        else Seq.empty[PathRef]
      res
    }

    def sources = Seq.empty[PathRef]

    def moduleDeps = mods(os.list(fileSrc / crossValue))
      .map(path =>
        millOuterCtx.enclosingModule.asInstanceOf[Cross[TreeCross]](path)
      )
  }

  def sources = Seq.empty[PathRef]
  def resources = Seq.empty[PathRef]
  def compileResources = Seq.empty[PathRef]

  def moduleDeps = super.moduleDeps ++ millModuleDirectChildren
    .collect { case m: Cross[_] => m }
    .flatMap(
      _.items
        .filter(_.crossSegments.length == 1)
        .map(_.module.value)
        .asInstanceOf[Seq[TreeCross]]
    )
}
```

**Reply by lolgab** [reply]
You can go another step further and make the modules intransitive, so you can only access the code in the current layer and the immediate children. This can be used to enforce a layered structure where Layer N can only access code in Layer N+1 but not N+2 and beyond.
Of course to share data structures you need some shared data, but that can be added to a separate normal module which is then added to `moduleDeps`.
To make a module intransitive you need to add a trait like:
```scala

trait IntransitiveModule extends JavaModule {
  override def bspTransitiveCompileClasspath: T[Agg[UnresolvedPath]] = T {
    T.traverse(moduleDeps)(m =>
      T.task {
        m.localCompileClasspath()
          .map(p => UnresolvedPath.ResolvedPath(p.path)) ++
          Agg(m.bspCompileClassesPath())
      }
    )()
      .flatten
  }
  override def transitiveCompileClasspath: T[Agg[PathRef]] = T {
    T.traverse(moduleDeps)(m =>
      T.task { m.localCompileClasspath() ++ Agg(m.compile().classes) }
    )()
      .flatten
  }
}
```
and mix it in `TreeModule` and `TreeCross`.

**Reply by haemin-tokenz** [reply]
@lolgab Thank you very much for the help.
For me the requirements are quite simple, so I have slightly tweaked your first version, and it works like a charm:

```scala
trait ParallelModule extends CommonModule { outer =>
  val fileSrc = millSourcePath / "src"
  def fileSrcInput = T.input { PathRef(fileSrc) }
  interp.watch(fileSrc)

  def inputModuleDeps: Seq[JavaModule]

  val submoduleFiles = os
    .walk(fileSrc)
    .filter(f => f.ext == "scala" || f.ext == "java")
    .map(_.relativeTo(fileSrc))

  def allSourceFiles = T { List[PathRef]() }

  object file extends Cross[CrossFile](submoduleFiles.map(_.toString.replace("/", "_")))
  trait CrossFile extends CommonModule with Cross.Module[String] {
    def allSourceFiles = T { List(PathRef(fileSrc / crossValue)) }
    def sources = Seq.empty[PathRef]
    def moduleDeps = inputModuleDeps
  }

  def moduleDeps = file.crossModules
}
```

```scala
  object implementation extends ParallelModule {
    def ivyDeps = super.ivyDeps() ++ Agg(
      // stuff
    )

    def inputModuleDeps = Seq(interface, definition)
  }
```

About the tree shaped dependencies -- interesting idea, but I wonder if linear dependency levels (0, 1, 2, 3, ...) using prefixes in the directory names `foo/src/00-shared-fragments` `foo/src/01-implementations` may be easier to deal with. Inside each level, none of the sources can see each other, but they can all see the ones above.

Obviously trees can express linear dependencies, but having to dig through the directory hierarchy may be a bit cumbersome. Personally my project structure isn't complex enough to require trees to describe, so just a thought :slightly_smiling_face:

**Reply by haemin-tokenz** [reply]
Sad news. Previously, it took 30 seconds to compile the whole thing single threaded. 
After I parallelized the module compilation, the longest file took 24 seconds (there 30+ files right now).

In my machine, parallelization makes compilation take more or less the same (wall clock), and requires more cpu and ram, so I am going to revert this change.

**Reply by lolgab** [reply]
You can try to split in less groups, maybe 2 groups and see if you see some benefit.

**Reply by bishabosha** [reply]
each time you compile that makes a separate time the compiler has to initialise the entire classpath. It would be much more efficient for the compiler itself to become batch-aware

**Reply by haemin-tokenz** [reply]
I see. This module sits at the near bottom of the module hierarchy, so the visible classpath is quite big. Also, each file of the module contains Scala 3 code that uses a lot of macros (quill). 

I don't know which one contributes more to the initialization overhead. Let me try to fiddle with the granularity and see if I can find an optimal level of parallelism.

If only there is a way to pre-heat and reuse the compiler state, or if the compiler is to be batch aware.. (but then this is not a mill issue anymore)
